README.md

```diff
@@ -1,2 +1,63 @@
+
+## Challenges
+
+### Which datastore is the right choice (Postgres vs Mongodb)?
+
+the service is write heavy, so I need a datastore that can scale and adapt to the scale with time.
+
+* My first approach was to choose Postgres ( I haven't considered Cassandra), given the [benchmarks](https://www.enterprisedb.com/news/new-benchmarks-show-postgres-dominating-mongodb-varied-workloads), PG beats MongoDB across writes and reads with more solid features when it comes to native support for transactions and being fully ACID complaint, the major drawback is it does not support [sharding natively](https://postgrespro.com/blog/pgsql/5969681), while it supports partitioning but its a master only write so keeping it fault tolerant is [tricky](https://www.alibabacloud.com/blog/solution-to-ensure-availability-and-reliability-rporto-of-postgresql-two-node-stream-replication_598027).
+
+* My choice depends on MongoDB, it offers sharding with write capability on the sharded replica, that way I can scale up the ingestion nodes as the number of events grow without worrying about the datastore not scaling accordingly, using [Hashed Sharding](https://www.mongodb.com/docs/manual/core/hashed-sharding/).
+the drawbacks is the load balancer needs to recognize and route events based on the sharded key implementation), also using [PSS](https://www.mongodb.com/docs/manual/core/replica-set-architecture-three-members/#primary-with-two-secondary-members--p-s-s-) replication for each shard to ensure fault tolerance.
+The way the keys are sharded early on will make huge difference on how the system will scale.
+
+* deploying two sets of the service, a set responsible for ingestion and other set for reading domain status.
+![Simple Diagram!](flow.png)
+
+
+### optimizations
+
+* using [hashed index](https://www.mongodb.com/docs/manual/indexes/#hashed-indexes), keys will be domain names, compatible with Hashed sharding.
+* since status can be NRT, the ingestion service can fan in a number of events in an internal queue then store them in a single [transaction](https://www.mongodb.com/docs/manual/core/transactions/).
+* caching the domain status once its `not catch-all` in a near memory cache like redis can speed up domain checks.
+
+### How can we make this better?
+
+#### events
+
+* event types, event name should be explicit enough, adding a type does not tell me what is this event unless I open it, parse it then do whatever logic.
+* events should carry at least one timestamp representing when this event was sourced.
+* events should carry one unique key to ensure integrity, its hard for the service to determine if it processed this event before or not.
+
+#### medium
+
+* Restful is bad for streaming its built on top of HTTP/1.1, I'd say for a service like this lets use GRPC it supports streaming and offers great multiplexing features
```

Comment:
instead of `lets use GRPC`
- we need to have less payload size, json is not the best payload format. we can rely on Thrift, Avro, or protobuff
- instead of HTTP as network protocol we can use rpc calls (for ex. gRPC) or sockets, or UDP